text/0000-eliminate-finalize.md

```diff
@@ -0,0 +1,142 @@
+
+- Title: Eliminate Finalize Step
+- Authors: [David Burkett](mailto:davidburkett38@gmail.com)
+- Start date: Aug 09, 2020
+- RFC PR: Edit if merged: [mimblewimble/grin-rfcs#0000](https://github.com/mimblewimble/grin-rfcs/pull/0000) 
+- Tracking issue: [Edit if merged with link to tracking github issue]
+
+---
+
+# Summary
+[summary]: #summary
+
+Use one-time addresses which include a pre-generated public excess and public nonce in order to eliminate the 3rd step in transaction building (finalize) [1].
+
+# Motivation
+[motivation]: #motivation
+
+To simplify transaction building.
+
+# Community-level explanation
+[community-level-explanation]: #community-level-explanation
+
+This feels like unnecessary red tape. See summary.
```

Comment:
I think it would be useful to outline high level `before` and `after` flows for transaction building, and how it might be experienced from the end user perspective. Perhaps something like the below? 

## Prior to this RFC

### Sender initiated (regular workflow)
1. Sender creates and sends partial slate to receiver to a slatepack address.
1. Receiver produces a response and returns to sender.
1. Sender finalizes and broadcasts to the network.

### Receiver initiated (invoice workflow)
1. Receiver creates a payment request and partial slate and sends to sender.
1. Sender approves the request, produces a response and returns to receiver.
1. Receiver finalises and broadcasts to the network.

## Flow as per this RFC

### Sender initiated (regular workflow)
1. Receiver shares a one-time address with sender along with `grin1` slatepack address.
1. Sender creates a partial slate and shares with receiver.
1. Receiver completes the slate and broadcasts transaction to the network.

### Receiver initiated (invoice workflow)
1. Receiver sends payment request, one-time address, and partial slate to sender.
1. Sender approves the request, completes the slate, and broadcasts transaction to the network.

Reply:
I will make a few small fixes to this, and use it as the community-level explanation. Thank you

Reply:
> ## Flow as per this RFC
> ### Sender initiated (regular workflow)
> 1. Receiver shares a one-time address with sender along with `grin1` slatepack address.

To me that's receiver initiated.
Whoever first sends the first slate contents, i.e. public excess and nonce, is the initiator.
Isn't one-time address is just a misleading name for initial slate contents?!
So this looks rather identical to 1st step of traditional receiver workflow "Receiver creates a payment request and partial slate and sends to sender."

> 2. Sender creates a partial slate and shares with receiver.

That looks identical to 2nd step of traditional receiver workflow "Sender approves the request, produces a response and returns to receiver."

> 3. Receiver completes the slate and broadcasts transaction to the network.

That looks identical to the 3rd step of traditional receiver workflow "Receiver finalizes and broadcasts to the network."

The only interesting part to me is the ability to still have a payment proof in the receiver workflow.

> ### Receiver initiated (invoice workflow)
> 1. Receiver sends payment request, one-time address, and partial slate to sender.
> 2. Sender approves the request, completes the slate, and broadcasts transaction to the network.

I fail to see how this works, given that the receiver never produced a partial signature.

Reply:
> Whoever first sends the first slate contents, i.e. public excess and nonce, is the initiator.
> Isn't one-time address is just a misleading name for initial slate contents?!

I believe you're thinking about this too technically, and not enough like a typical user. Right now, nobody complains about calling the regular workflow "sender initiated", even though the first part includes receiver sharing the address with the sender. And the address could just as easily be considered part of the slate as the one-time public nonce and excess are. If you think of the pub excess, pub nonce, and slatepack address as a "one-time address", and not as a partial slate, the RFC wording will make much more sense. 

> I fail to see how this works, given that the receiver never produced a partial signature.

You're correct, this part was copied from Daniel's example and modified in haste. It is indeed inaccurate. I'll fix this.

Reply:
Isn't the point of this RFC to get rid of the third step for both sender and receiver initiated transactions?

> ## Flow as per this RFC
> ### Sender initiated (regular workflow)
> 1. Receiver shares a one-time address with sender along with `grin1` slatepack address.
> 2. Sender creates a partial slate and shares with receiver.
> 3. Receiver completes the slate and broadcasts transaction to the network.
> 
> ### Receiver initiated (invoice workflow)
> 1. Receiver sends payment request, one-time address, and partial slate to sender.
> 2. Sender approves the request, completes the slate, and broadcasts transaction to the network.

Reply:
Yep, my proposal is here since one month: https://forum.grin.mw/t/eliminating-finalize-step/7621/22?u=kurt (add R_a in the hash of x, plus also verify nonce uniqueness) with some detailed security analysis on various aspects here: https://forum.grin.mw/t/eliminating-finalize-step/7621/40?u=kurt, here: https://forum.grin.mw/t/eliminating-finalize-step/7621/44?u=kurt, here: https://forum.grin.mw/t/eliminating-finalize-step/7621/78?u=kurt, here: https://forum.grin.mw/t/eliminating-finalize-step/7621/88?u=kurt and here: https://forum.grin.mw/t/integrated-payment-proofs-and-round-minimization/7745/7?u=kurt

Reply:
That's what I thought, thanks for clarifying. Wasn't sure if something changed after reading this.